### PR TITLE
Add power function and its unit tests.

### LINKED ISSUE #47: int_power_recursive challenge
## Description

Int Power Recursive Challenge
---

## Behavior

Returns the power of an integer base and an integer exponent.

---

### Example 1

Input: (1000, 0)
Output: 1

### Example 2

Input: (2003, 1)
Output: 2003

### Example 3

Input: (10, 3)
Output: 1000

### Example 4

Input: (10, 8)
Output: 100000000

### PR DESCRIPTION
**Pull request by nico1103**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/power.py
Comment (1sgtpepper) [inline solutions/power.py:1]: Filename could be more explicit regarding the behavior of the function instead of just power.py

File: solutions/tests/test_power.py
Comment (1sgtpepper) [inline solutions/tests/test_power.py:1]: Same suggestion with the filename.

File: solutions/power.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    Returns the power of an integer number.
+
+    Parameters:
+        base: int
+        exponent: int
+
+    Returns:
+        int
+
+    Raises:
```
Comment (Kuhai9801) [inline solutions/power.py:25]: This does not consider cases wherein the base is not 0, and the exponent is negative.

File: solutions/tests/test_power.py
Comment (1sgtpepper) [reply solutions/tests/test_power.py:1]: Consider the changes in the test case based on the changes that will be made from the feedback too!

File: solutions/power.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+def power(base, exponent):
+    """
+    Returns the power of an integer number.
+
+    Parameters:
+        base: int
+        exponent: int
+
+    Returns:
```
Comment (1sgtpepper) [inline solutions/power.py]: Does not consider cases wherein the exponent is negative, wherein the base is not 0 and the exponent is not -1. Ex. 9^-5 will return a float.

Comment (1sgtpepper) [reply solutions/power.py]: You may choose to ignore this, and change the description of your function instead.

File: solutions/tests/test_power.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    # Edge cases
+    def test_negative_base_exponent_zero(self):
+        """Exponent of 0 should always evaluate to 1."""
+        actual = power(-5000, 0)
+        expected = 1
+        self.assertEqual(actual, expected)
+
+    def test_base_one_any_exponent(self):
+        """Raising base 1 to any exponent should evaluate to 1."""
```
Comment (Kuhai9801) [inline solutions/tests/test_power.py:80]: There is lots of logic in unit tests such as calculations and comparisons.

File: solutions/power.py (hunk 2 above)
Comment (1sgtpepper) [reply solutions/power.py]: Assumptions are not clearly documented.

Comment (1sgtpepper) [reply solutions/power.py]: If you choose to document additional changes in your assumptions, make sure to change relevant parts of your code.

**Review by Kuhai9801 — CHANGES_REQUESTED** [review]
Great work overall! Some tweaking with the ambiguity of the filename, less logic, and documenting your assumptions well would make this even better.

**Comment by 1sgtpepper** [comment]
Also make sure you pass the py_formatting check. Refer to our Whatsapp group for the solution in case you are confused. Great Work!

File: solutions/tests/test_power.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        with self.assertRaises(TypeError):
+            power(2.2, 2.2)
+
+    def test_fractional_input(self):
+        """A fractional input should raise a type error."""
+        with self.assertRaises(TypeError):
+            power(8 / 2, 1 / 2)
+
+    def test_base_0_exponent_0(self):
+        """0 raised to the power of 0 is mathematically ambiguous."""
```
Comment (1sgtpepper) [inline solutions/tests/test_power.py]: Can you change the word ambiguous for more formality in mathematical terms?

File: solutions/tests/test_power.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+"""
+
+import unittest
+from ..power import power
+
+
+class TestPower(unittest.TestCase):
+    """Test the power function."""
+
+    # Standard Cases
```
Comment (1sgtpepper) [inline solutions/tests/test_power.py]: Redundant with lines 73-77

File: solutions/tests/test_power.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        self.assertAlmostEqual(actual, expected)
+
+    def test_negative_odd_exponent(self):
+        """The function should correctly with a negative odd exponent."""
+        actual = power(2, -3)
+        expected = 1 / 8
+        self.assertAlmostEqual(actual, expected)
+
+    # Edge cases
+    def test_negative_base_exponent_zero(self):
```
Comment (1sgtpepper) [inline solutions/tests/test_power.py]: redundant with lines 23-28

**Review by Kuhai9801 — CHANGES_REQUESTED** [review]
Made additional comments, wasn't able to spot these last time. My apologies, I was tired.

**Review by Kuhai9801 — APPROVED** [review]
Good job!
